server/service/async/async_host_seen.go

```diff
@@ -0,0 +1,190 @@
+package async
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+const (
+	hostSeenRecordedHostIDsKey   = "{host_seen:host_ids}"            // the SET of current (pending) host ids
+	hostSeenProcessingHostIDsKey = "{host_seen:host_ids}:processing" // the SET of host ids in the process of being collected
+	hostSeenKeysMinTTL           = 7 * 24 * time.Hour                // 1 week
+)
+
+// RecordHostLastSeen records that the specified host ID was seen.
+func (t *Task) RecordHostLastSeen(ctx context.Context, hostID uint) error {
+	if !t.AsyncEnabled {
+		t.seenHostSet.addHostID(hostID)
+		return nil
+	}
```

Comment:
I'd prefer the caller do this. It would be more readable and it's a bit unexpected that `async.Task.SomeMethod` could be doing it synchronously depending on config/creation-arg (same case with labels and policies).

Reply:
This way you also don't need to define "synchronized" related stuff on the `async` package (`// seenHostSet implements synchronized storage for the set of seen hosts.`)

Reply:
does this also apply to `FlushHostsLastSeen` below? I kind of like the idea of either completely separating "syncronized" vs "non-syncronized" code OR keeping all the `!task.AsyncEnabled` checks in a centralized place as much as possible (eg: moving the check that is done in `serve.go` here)

Reply:
I agree that it's not ideal API-wise, but it's quite useful as a way to have the full feature implemented in one location instead of scattered across files/packages. To be honest, I don't think it's really cleaner to have this condition check at the call-site. It's also how the other 2 async features are implemented, so for consistency I think it's best for all of them to take full "ownership" of the feature. Maybe the bigger issue is the naming instead of the code separation, though.

Reply:
(+1 to this being a naming issue, that was my first thought)

Reply:
> does this also apply to FlushHostsLastSeen below? I kind of like the idea of either completely separating "syncronized" vs "non-syncronized" code OR keeping all the !task.AsyncEnabled checks in a centralized place as much as possible (eg: moving the check that is done in serve.go here)

Yeah I'd like that too, although in that case I wanted to avoid creating a goroutine for nothing, hence the check inside serve.go. I don't know, it wouldn't be a big deal either to have it start and run a no-op every 1-10s, let me know how y'all feel about it, I don't mind either way!

Reply:
I'd trust your instinct! the code as-is LGTM!

Reply:
If we agree about the naming issue, I'll address this in a subsequent PR (and suggestions welcome, I think `task` or `job` is too generic, `asyncable`? it conveys that it can be but is not necessarily async?)

Reply:
+1 to naming change for now (And doing it on a separate PR sounds good too.)